query-algebrizer/src/types.rs

```diff
@@ -461,4 +464,140 @@ impl Debug for EmptyBecause {
             },
         }
     }
+}
+
+/// A set of `ValueType`s.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ValueTypeSet {
+    None,
+    Any,
+    One(ValueType),
+    Many(EnumSet<ValueType>),
```

Comment:
You could enforce your invariant statically and make this `Many(ValueType, EnumSet<ValueType>)`.  As it is, you need to comment on your invariant, and I think you should `assert` liberally that you never get `Many(empty())`.

I think this representation would make your cases cleaner almost everywhere.